### DISCUSSION TITLE
Is it possible to be a limit for china because of us-china trade war?

### DISCUSSION THREAD

**Discussion by SongOfYouth** [dotnet/csharplang#6201] [Q&A] [answered]
Some people thinking to turn to some else programing language like java in china, because they worried that the us-china trade war will make the C# became to be not available for chinese.

**Comment by richlander** [comment, marked as answer]
Microsoft has no policy on limiting the use of .NET for any reason. This topic literally never comes up.

Both Java and C# are sponsored by US companies and have open source implementations. In all likelihood, whatever happens (or doesn't happen) for one will be the same for the other. Also, there are other popular programming languages that match the pattern (sponsored by a US company and open source). No need to stop at C# and Java. I'm confident that nothing will happen.

Zooming out ... Trade wars typically use tariffs are their key implement. Both .NET and Java are free, so fly under the radar of tariffs. There's also bans, but no one wants to ban C#. Global developers (including in China) have and will always have unfettered access to C#. 

If Chinese developers want to learn more about C#, I'd recommend this Bilibili space: https://space.bilibili.com/483888821. My team has been contributing to it. Even better, the .NET Website has been translated to Chinese! Check it out: https://dotnet.microsoft.com/zh-cn/. Needless to say, the .NET Team is a friend of Chinese-language developers, wherever they may reside.

**Reply by SongOfYouth** [reply]
Thanks for your patient and useful reply!

**Reply by richlander** [reply]
Here's another good example of helping Chinese-language developers: https://github.com/dotnet/designs/pull/265#issuecomment-1146095672. Read from that comment until the resolution.

**Reply by NPUlrk** [reply]
Contradictory Responses and Practices. [Is .Net 6 subject to EAR? #74213](https://github.com/dotnet/runtime/discussions/74213)

**Reply by richlander** [reply]
That's just saying that .NET is subject to US export controls. I don't see anything contradictory.

**Reply by NPUlrk** [reply]
Microsoft has no policy on limiting the use of .NET for any reason. But they added open source .net to the EAR list after I raised this question. Joining the EAR means restricting the use, and more and more Chinese companies are being sanctioned by the US for trumped-up reason. I love C# and .net so much and use them every day at work. I am a little confused as to why they should cooperate with the US government as a weapon to fight against other countries when they are already open source.

**Reply by richlander** [reply]
I didn't research the situation. However, Microsoft has no choice but to cooperate with the US government with products it builds and distributes, particularly as it relates to rules like export controls. That doesn't affect the code on GitHub nor does it affect others who distributed open source .NET.

I should have added "except as required by law (US or otherwise)" to my original post. I apologize for not doing that.

It wasn't clear to me from that website what the effect of export controls actually is. I assume that can change over time.

Note: the reason that .NET is on that list may be due to cryptography APIs.

**Reply by NPUlrk** [reply]
Using technology that is added to the EAR list is always a risk for companies that comply strictly with the law. My company is a well-known multinational company. Many teams used C# and .net technologies, but they were forced to switch to Java or Python since the US sanctions. Experts with a voice say that C# and .Net belong to Microsoft and will affect the company's business continuity. We've been facing questions from all sides about the continued use of .Net. My intention is to get evidence here that .net is not controlled by EAR and Microsoft, but unfortunately I can only tell them that there are risks and that we may be able to compile .net ourselves once restricted. I hope this day doesn't come. I want technology to be truly borderless, and C# and net can benefit everyone, every corner of the planet.

**Reply by yaakov-h** [reply]
To my understanding Java (Oracle) and Python are also similarly subject to US export regulations. I don't think you're going to have much luck finding something that isn't...

**Reply by qrli** [reply]
@NPUlrk Not a lawyer but from what I know: US export control did expand to take opensource into the scene and requires opensource projects with encryption capability to register, thus even opensource project may have a code from EAR. However, because of the nature of opensource, it is not meaningful to ban the software. Instead, EAR defines subitems to differentiate the case which US really care about - which is software with encryption and especially non-standard encryption algorithm (which is rare). For majority of opensource projects, if there is any standard encryption algorithm included, they want to know it - so the project owner register and get an ECCN code, and that's it.

Practically speaking, all commercial software with ECCN 5D002 is at risk. However, for opensource, e.g. almost all projects under Apache are classified as 5D002, which is in a much worse position than .NET and many other opensource projects. Still, since they are opensource, they cannot really be banned from export in practice. At worst, US may demand commercial companies to stop providing them, but that will re-enforce the community stand on the software.

**Reply by qrli** [reply]
@NPUlrk Oracle Java is 5D992, same as .NET. (5D992 can be treat as safe.)
Python, even though not owned by a company, is even worse - it is classified as 5D002 because of encryption code. 
Even Linux kernel is 5D002. 
If you treat that as risk, you cannot use almost anything.

**Reply by NPUlrk** [reply]
@qrli The strategy I have now persuaded leaders and experts is that .net is the same 5D992 as the Windows operating system, so as long as the Windows operating system is available, .Net is available.

**Reply by qrli** [reply]
@NPUlrk  No, they are different. .NET is in a much better position than Windows.

Windows is a commercial software, which can be practically export-controlled regardless which classification EAR. They can change the rules anytime. The same applies to any commercial software from US, or even with components from US.

.NET is opensource. (It not only runs on Windows, but also Linux, BSD, etc.) The fact it is opensource has much bigger impact than the EAR classification.  In EAR, there are exception clauses specifically for opensource software as it is not practical to export-control. Even if the worst happened, 5D002 ones like Linux or Python will be the first to have trouble than .NET or Java.

**Reply by NPUlrk** [reply]
@qrli Thank you for your reply, it gave me a lot of confidence to continue using .net

**Comment by jamshedd** [comment]
I'm cross-posting my answer from another thread here so everyone here also has this available.

After discussions with our trade and legal teams we determined we _**can**_ classify all versions of modern .NET (.NET Core) as "Not subject to EAR". This change is now reflected on the Export Control website: https://www.microsoft.com/en-us/exporting/exporting-information.aspx

_Note: .NET **Framework** unlike modern .NET (NET Core) is a component of Windows and closed source, therefore all versions of .NET Framework will continue to be classified 5D992.c and are subject to EAR_
